include/cocaine/signal.hpp

```diff
@@ -0,0 +1,147 @@
+    along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+#ifndef COCAINE_SIGNAL_SET_HPP
+#define COCAINE_SIGNAL_SET_HPP
+
+#include <cocaine/locked_ptr.hpp>
+
+#include <atomic>
+#include <vector>
+#include <map>
+#include <functional>
+#include <signal.h>
+
+
+namespace cocaine { namespace signal {
+
+class cancellation_t {
+public:
+    // Call cancellation callback if present, was not called before and was not detached.
+    // Return true if callback was called
+    bool
+    cancel();
+
+    // Detach cancellation token so the destructor or cancel() would not call cancellation callback.
+    // Usually used right after async wait, f.e. signal_handler.async_wait(cb).detach();
+    // Return if callback was sucessfully detached(was not cancelled before, detached before and callback is present)
+    bool
+    detach();
+
+    cancellation_t();
+    cancellation_t(std::function<bool()> cancellation_callback);
+    cancellation_t(const cancellation_t&) = delete;
+    cancellation_t& operator=(const cancellation_t&) = delete;
+    cancellation_t(cancellation_t&&);
+    cancellation_t& operator=(cancellation_t&&);
+    ~cancellation_t();
+
+private:
+    std::atomic_flag cancelled;
+    std::function<bool()> cancellation_callback;
+};
+
+class handler_base_t {
+public:
+    enum class default_action_t {
+        IGNORE,           // Ignore signal
+        TERMINATE,        // Terminate via std::exit()
+        FORCED_TERMINATE, // Terminate via std::_Exit()
+        STOP_HANDLER      // Handler run method should return.
+    };
+
+    typedef std::function<void(const std::error_code&, int)> simple_callback_type;
+    typedef std::function<void(const std::error_code&, int, siginfo_t*)> detailed_callback_type;
+
+    virtual ~handler_base_t() {}
+
+    // Setup a single time handler for specified signal number.
+    // It should be reset up on signal.
+    // If you want to make recurrent one call async_wait inside handler in libasio manner
+    virtual cancellation_t
+    async_wait_simple(int signum, simple_callback_type handler) = 0;
+
+    // As the previous one but also pass siginfo_t to handler
+    virtual cancellation_t
+    async_wait(int signum, detailed_callback_type handler) = 0;
+
+    // If there are no handlers specified for a signal - handler runs default action.
+    // Here you can specify this default action.
+    virtual void
+    set_default_action(int signum, default_action_t action) = 0;
+};
+
+class handler_t : public handler_base_t {
+public:
+    typedef handler_base_t::default_action_t default_action_t;
+    typedef handler_base_t::simple_callback_type simple_callback_type;
+    typedef handler_base_t::detailed_callback_type detailed_callback_type;
+
+    typedef std::map<uint64_t, detailed_callback_type> detailed_callback_storage;
+
+    handler_t();
+
+    handler_t(const handler_t&) = delete;
+    handler_t(handler_t&&) = delete;
```

Comment:
it's not safe to move running handler. So looks like it's better to disallow this.